src/violation.rs

```diff
@@ -1,74 +1,17 @@
-/// Represents Violations found during validation of values based on schema types
-pub trait Violations {
-    /// adds a violation into the collection of [Violation]s
-    fn add_violation(&mut self, violation: ViolationImpl);
+use std::fmt;
+use std::fmt::Formatter;
+use thiserror::Error;
 
-    /// Returns `true` if no violations were found, otherwise `false`
-    fn is_valid(&self) -> bool;
-
-    /// Returns the [Violation]s that were found during validation
-    fn violations(&self) -> &[ViolationImpl];
-
-    /// Creates a checkpoint to track the number of violations from this point onwards
-    fn checkpoint(&self) -> Checkpoint;
-}
-
-// Represents collection of [Violation]s
-pub struct ViolationsImpl {
-    violations: Vec<ViolationImpl>,
-}
-
-impl ViolationsImpl {
-    pub fn new() -> Self {
-        Self { violations: vec![] }
-    }
-}
-
-impl Violations for ViolationsImpl {
-    fn add_violation(&mut self, violation: ViolationImpl) {
-        self.violations.push(violation);
-    }
-
-    fn is_valid(&self) -> bool {
-        self.violations.is_empty()
-    }
-
-    fn violations(&self) -> &[ViolationImpl] {
-        &self.violations
-    }
-
-    fn checkpoint(&self) -> Checkpoint {
-        Checkpoint::new(self.violations.len())
-    }
+/// Represents [Violation] found during validation with detailed error message, error code and the constraint for which the validation failed
+#[derive(Debug, Clone, Error)]
+pub struct Violation {
+    constraint: String, // represents the constraint that created this violation
+    code: String,       // represents an error code that indicates the type of the violation
```

Comment:
Are we copying these error code definitions from `ion-schema-kotlin`? I can see why they're useful, but I wonder if they should be formally defined instead of just a convention.

Reply:
Yes, I have used `ion-schema-kotlin` for reference to these error codes. But I will create an issue to make this an enum which defines all the error codes to be used by a `Violation`.